src/crypto/mod.rs

```diff
@@ -156,47 +170,107 @@ impl fmt::Debug for SignatureShare {
     }
 }
 
-/// A secret key.
-#[derive(Clone, PartialEq, Eq, Rand)]
-pub struct SecretKey(Fr);
+/// A secret key; wraps a single prime field element. The field element is
+/// heap allocated to avoid any stack copying that results when passing
+/// `SecretKey`'s around as arguments or as return values.
+#[derive(PartialEq, Eq)]
+pub struct SecretKey(Box<Fr>);
 
-impl fmt::Debug for SecretKey {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let uncomp = self.public_key().0.into_affine().into_uncompressed();
-        let bytes = uncomp.as_ref();
-        write!(f, "SecretKey({:?})", HexBytes(bytes))
+impl Default for SecretKey {
+    fn default() -> Self {
+        let mut fr = Fr::zero();
+        SecretKey::from_mut_ptr(&mut fr as *mut Fr)
     }
 }
 
-impl Default for SecretKey {
-    fn default() -> Self {
-        SecretKey(Fr::zero())
+impl Rand for SecretKey {
+    fn rand<R: Rng>(rng: &mut R) -> Self {
+        let mut fr = Fr::rand(rng);
+        SecretKey::from_mut_ptr(&mut fr as *mut Fr)
+    }
+}
+
+impl Clone for SecretKey {
+    fn clone(&self) -> Self {
+        let mut fr = *self.0;
+        SecretKey::from_mut_ptr(&mut fr as *mut Fr)
     }
 }
 
 impl Drop for SecretKey {
     fn drop(&mut self) {
-        let ptr = self as *mut Self;
+        self.zero_secret_memory();
+        self.munlock_secret_memory();
+    }
+}
+
+/// A debug statement where the secret prime field element is redacted.
+impl fmt::Debug for SecretKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "SecretKey(...)")
+    }
+}
+
+impl ContainsSecret for SecretKey {
+    fn mlock_secret_memory(&self) {
+        let ptr = &*self.0 as *const Fr as *mut u8;
+        let n_bytes = size_of_val(&*self.0);
+        unsafe {
+            if !mlock(ptr, n_bytes) {
+                println!("SK MLOCK FAILED!");
```

Comment:
Could this be a `panic!`? Or an `error!`? Or a `Result<(), errno::Errno>`? Arguably a `panic!` would be too strong. 

A `Result` would be better, in my opinion. Then all callers would need to handle the `Result`. Currently there is no special handling for situations such as when `Poly::from` is unable to `mlock` secret memory. Returning and handling the `Result` would allow this handling to take place.

However it would be perfectly fine to just log using `error!` if all you eventually do to handle the result would just be logging.

In Unix, `mlock` has many possible error codes. `memsec::mlock` returns a simple `bool` but the error code might still be recoverable using https://docs.rs/errno/0.2.4/errno/. If so, the panic or error message can include the description of the error, or at least the numeric error code.